events.py

```diff
@@ -0,0 +1,181 @@
+import threading
+
+
+class UsageError(Exception):
+    """
+    This is raised when you're trying to do something with an event that
+    isn't allowed.
+    """
+    pass
+
+
+class _ContextManagerMixin(object):  # Inherit object for Python2 compatibility
+    """
+    Inherit from this for Events that are created within a context manager.
+    """
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, exception_traceback):
+        # Ignore any exceptions raised; they'll get reraised outside our
+        # context anyway. Just clean up our references.
+        self.destruct()
+
+    def destruct(self):  # Call this when you want the Event to go out of scope.
+        pass
+
+
+class Event(_ContextManagerMixin):
+    """
+    You can pass these into `AnyEvent` or `AllEvent` below to join
+    these together. Beyond that, they act like `threading.Event` objects.
+    """
+    def __init__(self):
+        self._event = threading.Event()
+        # We have a lock to make set() and clear() atomic. It is re-entrant so
+        # that we can create a callback that can atomically check various
+        # conditions before running set or clear, without deadlocking (used in
+        # AnyEvent and AllEvent, below).
+        self._lock = threading.RLock()
```

Comment:
With so much use of the lock around the whole block of code within a function, should we put in the lock decorator currently residing in `pylib/config.py` and use that (but with hidden `_lock`)?

Reply:
Good idea! but I'm inclined to make that a separate PR.